shell/platform/embedder/embedder.cc

```diff
@@ -532,6 +533,70 @@ struct _FlutterPlatformMessageResponseHandle {
   fml::RefPtr<flutter::PlatformMessage> message;
 };
 
+struct LoadedELFDeleter {
+  void operator()(Dart_LoadedElf* elf) { ::Dart_UnloadELF(elf); }
+};
+
+using UniqueLoadedELF = std::unique_ptr<Dart_LoadedElf, LoadedELFDeleter>;
+
+struct _FlutterEngineAOTData {
+  UniqueLoadedELF loaded_elf = nullptr;
+  const uint8_t* vm_snapshot_data = nullptr;
+  const uint8_t* vm_snapshot_instrs = nullptr;
+  const uint8_t* vm_isolate_data = nullptr;
+  const uint8_t* vm_isolate_instrs = nullptr;
+};
+
+FlutterEngineResult FlutterEngineCreateAOTData(
+    const FlutterEngineAOTDataSource* source,
+    FlutterEngineAOTData* data_out) {
+  switch (source->type) {
+    case kFlutterEngineAOTDataSourceTypeElfPath: {
+      if (!source->elf_path || !fml::IsFile(source->elf_path)) {
+        return LOG_EMBEDDER_ERROR(kInvalidArguments,
+                                  "Invalid ELF path specified.");
+      }
+
+      auto aot_data = new _FlutterEngineAOTData();
+      const char* error = nullptr;
+
+      auto loaded_elf = Dart_LoadELF(
+          source->elf_path,               // file path
+          0,                              // file offset
+          &error,                         // error (out)
+          &aot_data->vm_snapshot_data,    // vm snapshot data (out)
+          &aot_data->vm_snapshot_instrs,  // vm snapshot instr (out)
+          &aot_data->vm_isolate_data,     // vm isolate data (out)
+          &aot_data->vm_isolate_instrs    // vm isolate instr (out)
+      );
+
+      if (loaded_elf == nullptr) {
+        delete aot_data;
+        return LOG_EMBEDDER_ERROR(kInvalidArguments, error);
+      }
+
+      aot_data->loaded_elf.reset(loaded_elf);
+
+      *data_out = aot_data;
+    } break;
+    default:
+      return LOG_EMBEDDER_ERROR(
+          kInvalidArguments,
+          "Invalid FlutterEngineAOTDataSourceType type specified.");
```

Comment:
It would be better to pull this error handling outside of the `switch` and removing the `default` so that failing to update this code when adding a new type to the header is a compiler error.